### PR TITLE
hide_audit_columns policy

### LINKED ISSUE #2372: Limit audit log view for e.g. 1st level helpdesk
Audit log, only show selected columns in audit log and only show selected actions.

This could be helpful for 1st level helpdesk users. They e.g. might only want to take a look at actions like /validate/check and would not be interested in information like "administrator".

Via policy we could limit the view to certain columns and actions.

### PR DESCRIPTION
**Pull request by laclaro**
This PR introduces the hide_audit_columns policy which limits the returned audit columns of the sqlaudit

closes #2372

### CODE REVIEW

File: privacyidea/static/components/audit/views/audit.log.html [hunk 1]
```diff
@@ -22,77 +22,91 @@
         <table class="table table-bordered table-striped" id="tableauditlog">
             <thead>
             <tr>
-                <th translate>number</th>
-                <th class="pifilter"><translate>date</translate>
+                <th ng-hide="!(auditdata.auditdata[0].hasOwnProperty('number'))"><translate>number</translate></th>
```
Comment (cornelinux) [inline privacyidea/static/components/audit/views/audit.log.html:25]: Could it ever happen, that we have no entry in the list ``auditdata.auditdata``?
Then we would get a javascript error here.

Another possiblity would be to add the information about the hidden columns as a fixed information here:
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/lib/audit.py#L114
in a static value like

    "hidden_columns": ["number", "sig_data"]

then we could simply check for the existance in the list:

    <th ng-hide=auditdata.hidden_columns.includes("number) translate>number</th>

(Also: You do not need to add an extra tag ``translate``. Simply put it as aparameter in the ``th`` tag.

This also applies to the other occurrences.

**Review by cornelinux — CHANGES_REQUESTED** [review]
* I recommend removing the hiding from the SQL level.

Please add tests in the policy lib level:

* Add prepolicy api lib tests around here:
  https://github.com/privacyidea/privacyidea/blob/master/tests/test_api_lib_policy.py#L1470

* Add REST API tests around here:
  https://github.com/privacyidea/privacyidea/blob/master/tests/test_api_audit.py

File: privacyidea/api/lib/prepolicy.py [hunk 2]
```diff
@@ -898,6 +898,29 @@ def auditlog_age(request=None, action=None):
...
+
+    :param request: The request that is intercepted during the API call
+    :type request: Request Object
+    :param action: An optional Action
+    :type action: basestring
+    :returns: Always true. Modified the parameter request
+    """
+    hidden_columns = Match.admin_or_user(g, action=ACTION.HIDE_AUDIT_COLUMNS,
+                                         user_obj=request.User).action_values(unique=False)
+    request.all_data["hidden_columns"] = list(hidden_columns)
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/api/lib/prepolicy.py:919]: No, I am wrong. action_values returns a dictionary. And you are right, you get the values of all policies by converting to a list (the keys to a list). So this is totally fine.

**Review by laclaro — COMMENTED** [review]
I changed the policy to only act on the API and lib level for the /audit endpoint (search method).

The policy does not restrict the columns obtained in csv download.

The WebUI now checks the available columns and hides all columns which are not available in the response (may it be by hide_audit_columns or anything else).

The commits are squashed and organized in PR #2500

**Comment by laclaro** [comment]
I changed the policy to only act on the API and lib level for the /audit endpoint (search method).

The policy does not restrict the columns obtained in csv download.

The WebUI now checks the available columns and hides all columns which are not available in the response (may it be by hide_audit_columns or anything else).

The commits are squashed and organized in PR #2500

File: privacyidea/static/components/audit/views/audit.log.html (hunk 1 above)
Comment (laclaro) [reply privacyidea/static/components/audit/views/audit.log.html]: done in #2500 commit https://github.com/privacyidea/privacyidea/pull/2500/commits/1a81fded8e075b728d2485b0851ce2c553cd3e87

File: privacyidea/lib/auditmodules/sqlaudit.py [hunk 3]
```diff
@@ -360,7 +360,7 @@ def _get_logentry_attribute(key):
                     'clearance_level': LogEntry.clearance_level}
         return sortname.get(key)
 
-    def csv_generator(self, param=None, user=None, timelimit=None):
+    def csv_generator(self, param=None, user=None, timelimit=None, hidden_columns=None):
```
[the remark this answers is not in the document]
Comment (laclaro) [reply privacyidea/lib/auditmodules/sqlaudit.py]: We decided to leave download csv untouched by this policy. A note was added in the docs #2500 commit https://github.com/privacyidea/privacyidea/pull/2500/commits/0fd9157d5d2486081a4ebc2ea042fc361f18aa86

File: privacyidea/lib/auditmodules/sqlaudit.py [hunk 4]
```diff
@@ -485,6 +488,12 @@ def clear(self):
         self.session.query(LogEntry).delete()
         self.session.commit()
 
+    def remove_hidden_columns(self, audit_dict, hidden_columns):
```
[the remark this answers is not in the document]
Comment (laclaro) [reply privacyidea/lib/auditmodules/sqlaudit.py]: removed

File: privacyidea/lib/auditmodules/sqlaudit.py [hunk 5]
```diff
@@ -395,7 +396,7 @@ def get_count(self, search_dict, timedelta=None, success=None):
         return log_count
 
     def search(self, search_dict, page_size=15, page=1, sortorder="asc",
-               timelimit=None):
+               timelimit=None, hidden_columns=None):
```
[the remark this answers is not in the document]
Comment (laclaro) [reply privacyidea/lib/auditmodules/sqlaudit.py]: done

File: doc/policies/admin.rst [hunk 6]
```diff
@@ -578,6 +578,14 @@ view older entries.
...
 Can be something like 10m (10 minutes), 10h (10 hours) or 10d (ten days).
 
+hide_audit_columns
+~~~~~~~~~~~~~~~~~~
+
+type: string
+
+With this admin policy the specified columns will not be displayed in the audit.
+For example a value ``sig_check log_level`` will hide these two columns.
+
```
[the remark this answers is not in the document]
Comment (laclaro) [reply doc/policies/admin.rst:589]: done in #2500 commit https://github.com/privacyidea/privacyidea/pull/2500/commits/0fd9157d5d2486081a4ebc2ea042fc361f18aa86

File: doc/policies/admin.rst [hunk 7]
```diff
@@ -578,6 +578,14 @@ view older entries.
 
 Can be something like 10m (10 minutes), 10h (10 hours) or 10d (ten days).
 
+hide_audit_columns
+~~~~~~~~~~~~~~~~~~
+
+type: string
+
+With this admin policy the specified columns will not be displayed in the audit.
```
[the remark this answers is not in the document]
Comment (laclaro) [reply doc/policies/admin.rst]: done

**Pull request closed without merging**